rflx/generator/common.py

```diff
@@ -310,6 +396,93 @@ def valid_path_to_next_field_condition(
+                    Selected("Ctx", "Buffer_First"),
+                    Selected("Ctx", "Buffer_Last"),
+                    Selected("Ctx", "First"),
+                    Name("Last"),
+                    Selected("Ctx", "Buffer"),
+                    Selected("Ctx", "Cursors"),
+                ),
+            ),
+            # WORKAROUND:
+            # Limitation of GNAT Community 2019 / SPARK Pro 20.0
+            # Provability of predicate is increased by adding part of
+            # predicate as assert
+            PragmaStatement(
+                "Assert", [str(self.message_structure_invariant(message, prefix=True))],
+            ),
+            Assignment(
+                Indexed(Selected("Ctx", "Cursors"), Name(field.affixed_name)),
+                NamedAggregate(
+                    ("State", Name("S_Structural_Valid")),
+                    ("First", Name("First")),
+                    ("Last", Name("Last")),
+                    ("Value", NamedAggregate(("Fld", Name(field.affixed_name))),),
+                    (
+                        "Predecessor",
+                        Selected(
+                            Indexed(Selected("Ctx", "Cursors"), Name(field.affixed_name),),
+                            "Predecessor",
+                        ),
+                    ),
+                ),
+            ),
+            Assignment(
+                Indexed(
+                    Selected("Ctx", "Cursors"),
+                    Call("Successor", [Name("Ctx"), Name(field.affixed_name)]),
+                ),
+                NamedAggregate(
+                    ("State", Name("S_Invalid")), ("Predecessor", Name(field.affixed_name)),
+                ),
+            ),
+        ]
+
+    def field_bit_location_declarations(self, field_name: Name) -> Sequence[Declaration]:
+        return [
+            ObjectDeclaration(
+                ["First"],
+                self.types.bit_index,
+                Call("Field_First", [Name("Ctx"), field_name]),
+                True,
+            ),
+            ObjectDeclaration(
+                ["Last"], self.types.bit_index, Call("Field_Last", [Name("Ctx"), field_name]), True,
+            ),
+        ]
+
+    def field_byte_location_declarations(self) -> Sequence[Declaration]:
+        return [
+            ExpressionFunctionDeclaration(
+                FunctionSpecification("Buffer_First", self.types.index),
+                Call(self.types.byte_index, [Name("First")]),
+            ),
+            ExpressionFunctionDeclaration(
+                FunctionSpecification("Buffer_Last", self.types.index),
+                Call(self.types.byte_index, [Name("Last")]),
+            ),
+            ExpressionFunctionDeclaration(
+                FunctionSpecification("Offset", self.types.offset),
+                Call(
+                    self.types.offset,
+                    [Mod(Sub(Number(8), Mod(Name("Last"), Number(8))), Number(8))],
+                ),
+            ),
+        ]
+
```

Comment:
Where does that code come from? I can't find a corresponding deletion that I'd expect from reformatting.

Reply:
This newly added code replaces code duplications. It is part of 63dd7fd.